core/domain/stats_jobs_one_off.py

```diff
@@ -18,14 +18,466 @@
+                    session_ids_set.add(state_hit_event['session_id'])
+                if state_hit_event['session_id'] not in (
+                        state_hit_events_session_id_mapping):
+                    state_hit_events_session_id_mapping[state_hit_event[
+                        'session_id']] = []
+                state_hit_events_session_id_mapping[state_hit_event[
+                    'session_id']].append(state_hit_event)
+
+            # Compute num_completions for the states.
+            for session_id in session_ids_set:
+                state_hit_event_dicts_sessioned = (
+                    state_hit_events_session_id_mapping[session_id])
+                # All state hit events in the above list are ordered by their
+                # timestamp. This implies that the last state hit event is the
+                # state at which the learner quit the exploration (This is only
+                # True if the last state hit event is not the terminal state).
+                # So, if the last state is not the terminal state, we increment
+                # num_completions for all states except the last state.
+                # Otherwise, we increment num_completions for all the states.
+                for state_hit_event in state_hit_event_dicts_sessioned[:-1]:
+                    state_stats_mapping[state_hit_event[
+                        'state_name']].num_completions_v1 += 1
+                if session_id in completed_session_ids:
+                    state_stats_mapping[state_hit_event_dicts_sessioned[-1][
+                        'state_name']].num_completions_v1 += 1
+
+            # Compute total_answers_count and useful_feedback_count.
+            for value in state_answer_events_for_this_version:
+                for answer in value['submitted_answer_list']:
+                    state_stats_mapping[value[
+                        'state_name']].total_answers_count_v1 += 1
+                    if answer['classification_categorization'] != (
+                            exp_domain.DEFAULT_OUTCOME_CLASSIFICATION):
+                        state_stats_mapping[value[
+                            'state_name']].useful_feedback_count_v1 += 1
+
+            # Now that the stats for the exploration's version are computed,
+            # calculate num_actual_starts. To compute this, we take the max of
+            # the first hit counts of all states that the initial state leads
+            # to. The max is taken instead of sum because the sum of
+            # first_hit_counts will sometimes end up counting a single learner
+            # twice. The max of both first hit counts ensures that we are
+            # counting only unique learners who traversed past the initial
+            # state.
+            max_first_hit_from_init_state = 0
+            for answer_group in versioned_exploration.states[versioned_exploration.init_state_name].interaction.answer_groups: # pylint: disable=line-too-long
```

Comment:
Pull versioned_exploration.states[versioned_exploration.init_state_name] into a separate variable named init_state. In general try to avoid pylint pragmas where possible.